src/%IPM/Main.cls

```diff
@@ -2383,11 +2405,178 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
 }
 
+ClassMethod EnableIPM(ByRef pCommandInfo)
+{
+	Set quiet = $$$HasModifier(pCommandInfo,"quiet")
+	Set globally = $$$HasModifier(pCommandInfo, "globally")
+	Set localOnly = $$$HasModifier(pCommandInfo, "localOnly")
+	Set version = $$$GetModifier(pCommandInfo, "version")
+	Set namespaces = $$$GetModifier(pCommandInfo, "namespaces")
+	Set useLocal = 1			// var to store the final decision of whether to use local manifest or get from server
+	Set targetVersion = "" 		// var to store the final version of IPM to be installed
+	Kill targetNamespaces 		// multi-dim array to store the final namespaces that need to install IPM
+
+	// Get list of local IPM artifacts under /<instance install dir>/lib/IPM/
+	Set XMLDir = $system.Util.InstallDirectory()
+	For subDirName = "lib", "IPM" {
+		Set XMLDir = ##class(%File).SubDirectoryName(XMLDir, subDirName)
+	}
+  	Set statement = ##class(%SQL.Statement).%New()
+  	Set sc = statement.%PrepareClassQuery("%File", "FileSet")
+  	If $$$ISERR(sc) {
+		Do $SYSTEM.Status.DisplayError(sc)
+		Quit
+	}
+	// Valid IPM installation manifest should be of format: ipm-0.0.1.xml
+	Set resultSet = statement.%Execute(XMLDir, "ipm-*.xml")
+	If ((resultSet.%SQLCODE '= 0) && '(resultSet.%Message [ "does not exist")) {
+		Write !, resultSet.%Message 
+		Quit
+	}
+	Kill ipmLocalArray			// multi-dim array to store local ipm manifests; ipmLocalArray(<version>)=<fileName>
+	Kill menuList				// menu to let user choose which version of local ipm to install (if not in quiet mode and no version is specified)
+	Set counter = 0
+	While resultSet.%Next(.sc) {
+		$$$ThrowOnError(sc)
+		Set fileName = resultSet.%Get("Name")
+		If (resultSet.%Get("Type") = "F") {
+			// get the exact version string from filename
+			Set ipmLocalArray($Piece($Piece(fileName, "ipm-", 2), ".xml")) = fileName
+			Set counter = counter + 1
+			Set menuList(counter) = $Piece($Piece(fileName, "ipm-", 2), ".xml")
+		}
+	}
+	$$$ThrowOnError(sc)
+
+	If (localOnly && (counter = 0)) {
+		$$$ThrowOnError($$$ERROR($$$GeneralError,"No ipm-<version>.xml installer file is found locally."))
+	}
+
+	If (version '= "") {
+		Set targetVersion = version
+		If '$DATA(ipmLocalArray(version)) {
+			If localOnly {
+				$$$ThrowOnError($$$ERROR($$$GeneralError,"No ipm-"_version_".xml installer file is found in "_XMLDir_" locally."))
+			} ElseIf 'quiet {
+				Write !, "No ipm-"_version_".xml installer file is found in "_XMLDir_" locally. Trying to download from the registry."
+			}
+			// need to get from remote registry server
+			Set useLocal = 0
+		}
+	} Else {
+		If quiet {
+			$$$ThrowOnError($$$ERROR($$$GeneralError,"Cannot install IPM in non-prompt mode without a specified IPM version."))
+		}
+		If (counter = 0) {
+			$$$ThrowOnError($$$ERROR($$$GeneralError,"No IPM version is found in "_XMLDir_" locally. Use version modifier to specify a version to download from the registry."))
+		}
+		// prompt user with local options
+		Set menuValue = 1
```

Comment:
The prompt should include local and remote versions for the user to choose from (need to accumulate available remote versions before hand)

Reply:
I'm somewhat hesitate to include remote versions in prompts, since they will be something look like ["0.0.4","0.0.5","0.0.6","0.0.7","0.1.0","0.1.1","0.1.2","0.1.3","0.1.4","0.1.5","0.1.6","0.1.7","0.1.8","0.2.1","0.2.10","0.2.11","0.2.12","0.2.13","0.2.14","0.2.2","0.2.3","0.2.4","0.2.5","0.2.6","0.2.7","0.2.8","0.2.9","0.3.0","0.3.2","0.3.3","0.3.4"] which is a lot for any sort of prompt. We can either let the user input a number for remote version in prompt if they really want to, or force to specify one as modifier. Or, provide only a "latest" option. I'm more leaning towards the last actually, but open to discuss